File: node/beacon.go
package tsnode

// Hold beacons (PLAYLISTS.md §10, shipped): announce suppression deliberately hides
// holders (~one voice per playlist per cycle), so popularity is invisible. This side
// channel restores the signal cheaply: each client broadcasts a small jittered hourly
// manifest of truncated 8-byte name-hashes of its DISCLOSURE SET (held + published-mine
// — the same set the playlist-list protocol serves; never private, never seen-tier).
// Every node counts hash → distinct origins over a 24h sliding window, locally and
// independently — no consensus, none needed. Gossipsub's default StrictSign
// authenticates the origin for free; Sybil-gameability is accepted until the identity
// layer lands (only the WEIGHTING of this feed changes then, never the feed itself).

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"math/rand"
	"sync"
	"time"

	lru "github.com/hashicorp/golang-lru/v2"
	pubsub "github.com/libp2p/go-libp2p-pubsub"
	"github.com/libp2p/go-libp2p/core/peer"
	"golang.org/x/time/rate"
)

const (
	beaconVersion byte = 0x01
	// beaconMaxHashes bounds one beacon message: honest libraries are hundreds of
	// playlists; 4096 × 8 B ≈ 32 KiB is already implausibly large. Bigger ⇒ Reject.
	beaconMaxHashes = 4096
	beaconHashCap   = 16384 // outer LRU: distinct playlist hashes we count for
	// beaconOriginsPerHash caps the per-playlist origin set — the count SATURATES
	// there ("~512+ holders" needs no more precision, and a Sybil can't bloat it).
	beaconOriginsPerHash = 512
	beaconPeerLims       = 4096 // bounded per-peer limiter table (fwd.go doctrine)
)

// Tunables kept as vars so tests can shrink them.
var (
	beaconInterval   = 60 * time.Minute // base publish cadence
	beaconJitter     = 15 * time.Minute // ± uniform jitter on the cadence
	beaconFirstDelay = 60 * time.Second // first beacon 1–2 min after a manifest exists
	beaconKickMin    = 10 * time.Minute // manifest-change kick floor (else wait for cycle)
	beaconWindow     = 24 * time.Hour   // sliding count window
	// beaconOriginMinGap Ignores repeat beacons from one origin: an honest node beacons
	// hourly; anything under this is either a kick burst or a count-inflation attempt.
	beaconOriginMinGap = 2 * time.Minute
	// First-hop rate caps (charged to the peer we received from, own publishes exempt).
	beaconPeerMsgRate   = rate.Limit(0.1) // sustained msgs/s per peer
	beaconPeerMsgBurst  = 8
	beaconPeerByteRate  = rate.Limit(64 << 10) // sustained bytes/s per peer
	beaconPeerByteBurst = 256 << 10
)

// nameHash8 is the beacon identifier for a playlist name: sha256(name)[:8]. Truncation
// keeps beacons small; collisions just merge two counts, harmless for a ranking signal.
func nameHash8(name string) [8]byte {
	var h [8]byte
	s := sha256.Sum256([]byte(name))
	copy(h[:], s[:8])
	return h
}

// encodeBeacon frames [version][uvarint N][N × 8-byte hash].
func encodeBeacon(hashes [][8]byte) []byte {
	buf := make([]byte, 0, 1+binary.MaxVarintLen64+len(hashes)*8)
	buf = append(buf, beaconVersion)
	buf = binary.AppendUvarint(buf, uint64(len(hashes)))
	for _, h := range hashes {
		buf = append(buf, h[:]...)
	}
	return buf
}

// decodeBeacon rejects anything but an exact-length, cap-respecting frame.
func decodeBeacon(b []byte) ([][8]byte, error) {
	if len(b) == 0 || b[0] != beaconVersion {
		return nil, fmt.Errorf("bad beacon version")
	}
	count, n := binary.Uvarint(b[1:])
	if n <= 0 || count == 0 || count > beaconMaxHashes {
		return nil, fmt.Errorf("bad beacon count")
	}
	body := b[1+n:]
	if uint64(len(body)) != count*8 {
		return nil, fmt.Errorf("beacon length mismatch")
	}
	hashes := make([][8]byte, count)
	for i := range hashes {
		copy(hashes[i][:], body[i*8:])
	}
	return hashes, nil
}

// originSet is one playlist-hash's holder ledger: origin → last beacon time, capped
// (count saturates) and pruned to the window on every touch/read.
type originSet struct {
	mu sync.Mutex
	m  map[peer.ID]time.Time
}

func (o *originSet) touch(p peer.ID, now time.Time) {
	o.mu.Lock()
	defer o.mu.Unlock()
	for id, at := range o.m {
		if now.Sub(at) > beaconWindow {
			delete(o.m, id)
		}
	}
	if _, ok := o.m[p]; !ok && len(o.m) >= beaconOriginsPerHash {
		return // saturated
	}
	o.m[p] = now
}

func (o *originSet) count(now time.Time) int {
	o.mu.Lock()
	defer o.mu.Unlock()
	c := 0
	for _, at := range o.m {
		if now.Sub(at) <= beaconWindow {
			c++
		}
	}
	return c
}

// beaconState holds the counting store + the serving-side rate tables + publish clock.
type beaconState struct {
	counts   *lru.Cache[[8]byte, *originSet]
	origLast *lru.Cache[peer.ID, time.Time] // per-ORIGIN min-gap ledger (count integrity)
	lims     *lru.Cache[peer.ID, *plLimiter] // per-FIRST-HOP rate buckets (flood cap)

	mu      sync.Mutex
	lastPub time.Time
}

func newBeaconState() *beaconState {
	return &beaconState{
		counts:   mustLRU[[8]byte, *originSet](beaconHashCap),
		origLast: mustLRU[peer.ID, time.Time](beaconHashCap),
		lims:     mustLRU[peer.ID, *plLimiter](beaconPeerLims),
	}
}

func (s *beaconState) allow(from peer.ID, size int) bool {
	l, ok := s.lims.Get(from)
	if !ok {
		l = &plLimiter{
			msgs:  rate.NewLimiter(beaconPeerMsgRate, beaconPeerMsgBurst),
			bytes: rate.NewLimiter(beaconPeerByteRate, beaconPeerByteBurst),
		}
		s.lims.Add(from, l)
	}
	now := time.Now()
	return l.msgs.AllowN(now, 1) && l.bytes.AllowN(now, size)
}

// record counts every hash for one authenticated origin.
func (s *beaconState) record(origin peer.ID, hashes [][8]byte) {
	now := time.Now()
	for _, h := range hashes {
		set, ok := s.counts.Get(h)
		if !ok {
			set = &originSet{m: map[peer.ID]time.Time{}}
			s.counts.Add(h, set)
		}
		set.touch(origin, now)
	}
}

// backers returns the current windowed holder count for each name (0 = unknown).
func (s *beaconState) backers(names []string) map[string]int {
	now := time.Now()
	out := make(map[string]int, len(names))
	for _, name := range names {
		c := 0
		if set, ok := s.counts.Get(nameHash8(name)); ok {
			c = set.count(now)
		}
		out[name] = c
	}
	return out
}

// beaconValidator gates the topic: first-hop rate (Ignore), format (Reject), and a
// per-origin minimum gap (Ignore — repeats can't inflate counts anyway, this just
// keeps them off the mesh). Own publishes exempt from rate policy.
func (n *Node) beaconValidator(_ context.Context, from peer.ID, m *pubsub.Message) pubsub.ValidationResult {
	if from != n.host.ID() && !n.beacons.allow(from, len(m.Data)) {
		return pubsub.ValidationIgnore
	}
	if _, err := decodeBeacon(m.Data); err != nil {
		return pubsub.ValidationReject
	}
	origin := m.GetFrom() // authenticated under gossipsub's default StrictSign
	if origin != n.host.ID() {
		if at, ok := n.beacons.origLast.Get(origin); ok && time.Since(at) < beaconOriginMinGap {
			return pubsub.ValidationIgnore
		}
		n.beacons.origLast.Add(origin, time.Now())
	}
	return pubsub.ValidationAccept
}

// beaconSink counts a validator-passed beacon for its origin.
func (n *Node) beaconSink(origin peer.ID, data []byte) {
	hashes, err := decodeBeacon(data)
	if err != nil {
		return // validator raced a tunable swap; just drop
	}
	n.beacons.record(origin, hashes)
}

// publishBeacon broadcasts the current disclosure-set manifest as hashes. No-op when
// the manifest is empty (nothing held ⇒ nothing to say).
func (n *Node) publishBeacon(ctx context.Context) {
	entries := n.plList.snapshot(beaconMaxHashes)
	if len(entries) == 0 {
		return
	}
	hashes := make([][8]byte, len(entries))
	for i, e := range entries {
		hashes[i] = nameHash8(e.Name)
	}
	if err := n.pubsub.PublishBeacon(ctx, encodeBeacon(hashes)); err != nil {
		n.logf("beacon publish failed: %v", err)
		return
	}
	n.beacons.mu.Lock()
	n.beacons.lastPub = time.Now()
	n.beacons.mu.Unlock()
}

// KickBeacon publishes ahead of the cycle after a manifest change, floor-limited so a
// hold/unhold flurry can't turn into a beacon flood. (The manifest RPC calls this; the
// desktop only posts manifests that actually changed.)
func (n *Node) KickBeacon(ctx context.Context) {
	n.beacons.mu.Lock()
	recent := time.Since(n.beacons.lastPub) < beaconKickMin
	n.beacons.mu.Unlock()
	if recent {
		return
	}
	n.publishBeacon(ctx)
}

// beaconLoop is the client's jittered hourly publish cadence. The first beacon goes
// out 1–2 minutes after the manifest first becomes non-empty (the mesh needs a moment,
// and so does the desktop's first manifest push).
func (n *Node) beaconLoop(ctx context.Context) {
	next := beaconFirstDelay + time.Duration(rand.Int63n(int64(beaconFirstDelay)))
	t := time.NewTimer(next)
	defer t.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-t.C:
		}
		n.beacons.mu.Lock()
		due := time.Since(n.beacons.lastPub) >= beaconKickMin // a kick may have just fired
		n.beacons.mu.Unlock()
		if due {
			n.publishBeacon(ctx)
		}
		jitter := time.Duration(rand.Int63n(int64(2*beaconJitter))) - beaconJitter
		t.Reset(beaconInterval + jitter)
	}
}

File: packages/playlists-web/src/beacon.ts
// Hold beacons — a port of node/beacon.go's wire format.
//
// A beacon says "I hold these playlists" without saying which peer holds what to anyone who wasn't
// listening: names are truncated to sha256(name)[:8]. Publishing it from the browser is ~30 lines
// and it makes web peers contribute to the backer counts DESKTOP users see.
import { sha256 } from "multiformats/hashes/sha2";

const BEACON_VERSION = 0x01;

/** sha256(name)[:8] — the beacon identifier for a playlist name. */
export async function nameHash8(name: string): Promise<Uint8Array> {
  const d = await sha256.digest(new TextEncoder().encode(name));
  return d.digest.subarray(0, 8);
}

/** `[version][uvarint N][N x 8-byte hash]` */
export function encodeBeacon(hashes: Uint8Array[]): Uint8Array {
  const head: number[] = [BEACON_VERSION];
  let n = hashes.length;
  while (n >= 0x80) {
    head.push((n & 0x7f) | 0x80);
    n >>>= 7;
  }
  head.push(n);
  const out = new Uint8Array(head.length + hashes.length * 8);
  out.set(head, 0);
  hashes.forEach((h, i) => out.set(h, head.length + i * 8));
  return out;
}

/** Signalled by decodeBeacon when the frame is a newer, unknown beacon version — IGNORE it (never
 *  penalise a peer that merely upgraded), the same forward-compat stance the playlist wire takes. */
export class UnknownBeaconVersion extends Error {}

/** Inverse of encodeBeacon — the counting half (node/beacon.go's decodeBeacon). Returns the 8-byte
 *  name hashes so the caller can tally distinct origins per playlist. Throws UnknownBeaconVersion on a
 *  future version and a plain Error on a structurally-malformed frame. Caps N so a hostile length can't
 *  make us allocate unboundedly. */
export function decodeBeacon(bytes: Uint8Array): Uint8Array[] {
  if (bytes.length < 1) throw new Error("beacon: empty");
  if (bytes[0] !== BEACON_VERSION) throw new UnknownBeaconVersion(`beacon version ${bytes[0]}`);
  let i = 1;
  let n = 0;
  let shift = 0;
  for (;;) {
    if (i >= bytes.length) throw new Error("beacon: truncated length");
    const b = bytes[i++];
    n |= (b & 0x7f) << shift;
    if ((b & 0x80) === 0) break;
    shift += 7;
    if (shift > 28) throw new Error("beacon: length overflow");
  }
  n >>>= 0;
  if (n > 4096) throw new Error("beacon: too many hashes");
  if (bytes.length - i !== n * 8) throw new Error("beacon: body length mismatch");
  const out: Uint8Array[] = [];
  for (let k = 0; k < n; k++) out.push(bytes.subarray(i + k * 8, i + k * 8 + 8));
  return out;
}

const BEACON_WINDOW_SECS = 24 * 60 * 60; // popularity window (matches beacon.go's 24h originSet prune)
const BEACON_MIN_GAP_SECS = 2 * 60; // per-origin min gap (beaconValidator's rate floor)
const BEACON_MAX_ORIGINS = 512; // per-name origin cap (originSet bound)

const hex = (b: Uint8Array): string => Array.from(b, (x) => x.toString(16).padStart(2, "0")).join("");

/** The 8-byte name hash as a hex key — the ledger keys and the backer lookup must agree on it. */
export const beaconHashHex = (h: Uint8Array): string => hex(h);

/** Distinct-origin backer counts over a 24h window — the counting half of the beacon protocol
 *  (node/beacon.go's originSet + backers). Keyed by the 8-byte name hash, so a playlist's popularity
 *  is countable without anyone disclosing which peer holds what. A per-origin min-gap (2 min) mirrors
 *  the Go beaconValidator, and the per-name origin table is capped so a flood can't grow it unbounded.
 *  `nowSecs` is passed in so the caller owns the clock. */
export class BeaconLedger {
  private byHash = new Map<string, Map<string, number>>(); // nameHashHex -> (origin -> lastSeenSecs)

  record(origin: string, hashes: Uint8Array[], nowSecs: number): void {
    for (const h of hashes) {
      const key = hex(h);
      let origins = this.byHash.get(key);
      if (!origins) {
        origins = new Map();
        this.byHash.set(key, origins);
      }
      const last = origins.get(origin);
      if (last !== undefined && nowSecs - last < BEACON_MIN_GAP_SECS) continue; // too soon from this origin
      origins.delete(origin); // re-insert to keep recent in the bounded eviction order
      if (origins.size >= BEACON_MAX_ORIGINS) origins.delete(origins.keys().next().value as string);
      origins.set(origin, nowSecs);
    }
  }

  count(nameHashHex: string, nowSecs: number): number {
    const origins = this.byHash.get(nameHashHex);
    if (!origins) return 0;
    let n = 0;
    for (const [o, t] of origins) {
      if (nowSecs - t < BEACON_WINDOW_SECS) n++;
      else origins.delete(o); // prune stale origins on read
    }
    return n;
  }
}
